Readme.md
# ddprof

The Datadog Native Profiler for Linux.

## Overview

`ddprof` is a command-line utility to gather profiling data. After install you will continuously see where your application is spending CPU and memory allocations.
The data will be available in the `/profiling` section of the [Datadog UI](https://app.datadoghq.com/).

## Quick Start

Our official documentation is available [here](https://docs.datadoghq.com/profiler/enabling/ddprof/?tab=environmentvariables).
Our pre-built binaries are compatible with both musl and glibc. You should not need to recompile `ddprof` from source. 

### From binary [Recommended]

An installation guide is available [here](https://docs.datadoghq.com/profiler/enabling/ddprof/?tab=environmentvariables).
Check out our Release page for our [latest](https://github.com/DataDog/ddprof/releases/tag/latest) release. Download the release and extract `ddprof`.
Instrumenting your application should be as simple as adding `ddprof` in front of your usual command line.

To install the profiler, check out our [installation-helpers](#Installation-helpers) bellow.

The following command will run `ddprof` with the default settings (CPU and allocations)

```bash
ddprof -S service_name_for_my_program ./my_program arg1 arg2
```

Profiling data shows up in the `/profiling` section of your Datadog UI. Specifying a service name will help you select your profiling data. 
Refer to [commands](docs/Commands.md) for a more advanced usage of `ddprof`.

### From source

Checkout our build section [here](./docs/Build.md).

## Prerequisites

### Perf event paranoid

The target machine must have `perf_event_paranoid` set to 2 or lower.

```bash
# needs to be less than or equal to 2
cat /proc/sys/kernel/perf_event_paranoid
```

Here is an example adding a startup configuration to your system. This requires a system restart.

```bash
sudo sh -c 'echo kernel.perf_event_paranoid=2 > /etc/sysctl.d/perf_event_paranoid_2.conf'
```

Alternatively you can use `CAP_SYS_ADMIN` or `sudo` as a one off test mechanism, more in the [Troubleshooting](./docs/Troubleshooting.md) section. 
Don't hesitate to [reach-out](#Reaching-out) if you are not able to use our profiler!

### Agent installation

It is recommended to have an agent setup on the system you are profiling.
By default the profiler will target `localhost:8126` (the default trace agent endpoint). The `DD_TRACE_AGENT_URL` environment variable can be used to override this setting.

## Installation helpers

### Ubuntu / Debian

The following commands will download and install `ddprof` on Debian or Ubuntu distributions:

```bash
export ARCH=$(dpkg --print-architecture) # ARCH should hold amd64 or arm64
# ddprof requires xz-utils to uncompress the archive
sudo apt-get update && \
sudo DEBIAN_FRONTEND=noninteractive apt-get install -y xz-utils curl jq && \
tag_name=$(curl -s https://api.github.com/repos/DataDog/ddprof/releases/latest | jq -r '.tag_name[1:]') && \
url_release="https://github.com/DataDog/ddprof/releases/download/v${tag_name}/ddprof-${tag_name}-${ARCH}-linux.tar.xz" && \
curl -L -o ddprof-${ARCH}-linux.tar.xz ${url_release} && \
tar xvf ddprof-${ARCH}-linux.tar.xz && \
sudo mv ddprof/bin/ddprof /usr/local/bin && \
rm -Rf ddprof-amd64-linux.tar.xz ./ddprof && \
ddprof --version
```

## Key Features

### Simplicity

`ddprof` is a wrapper, so using it should be as simple as injecting the binary into your container and wrapping your entrypoint.
`ddprof` will use environment variables if they are available, overriding them with commandline parameters if given.

### Safety

- Minimal interference to execution of instrumented processes
- `ddprof`'s Memory usage is sandboxed

### Allocation profiling

- By working in user space, `ddprof` can instrument allocations with low overhead

## Docs

Architectural showpieces and such will always be available in the `docs/` folder.

- [Build](./docs/Build.md)
- [Design](./docs/Design.md)
- [Automatically updated list of commads](./docs/Commands.md)
- [Troubleshooting](./docs/Troubleshooting.md)

## Reaching-out

Any contribution is welcome. Please check out our guidelines [here](CONTRIBUTING.md).


changelog
0.11.0 to 0.11.1 (15/03/2023)
* Enable automatic releasing to GITHub

0.10.1 to 0.11.0 (13/03/2023)
* JITDump support
* Fix crash when using custom stack (example: Fibers)
* Fix tag validation

0.9.3 to 0.10.1 (14/11/2022)
* Allocation profiling - instrument forks
* Universal binary - remove dependency on libc
This allows compatibility with older libc (like centos 7)
* Formalize tracepoint grammar
* Add build-id to mapping information
* Use a temporary file to generate a library to load
* Switch user if failing to read /proc/<pid>/maps
* Changes in instrumentation mechanisms
Periodically check for new loaded libraries
Use GOT override instead of symbol interposition
* Add information on lost events
When ring buffer is full we flag them as "lost events".
* Use jemalloc as default allocator
This prevents a counter-performance on musl compilations

0.8.2 to 0.9.3 (08/08/2022)
* Allocation profiling
Library mode and wrapper mode have an option to instrument allocations with low overhead.
Limitation: This only works for dynamically linked allocators (example: standard allocator) and does not follow forks. If you are concerned by these limitations, please reach out with your requirements.
* Samples are labeled with PID / TID
* Preset option : adds defined presets to simplify configuration of the profiler (cpu_only, alloc_only)
* Catch kernel activity (when permissions allow it)
* Several improvements to unwinding
* Alpine musl (built on 3.16)

0.8.1 to 0.8.2 (14/04/2022)
* Correctly convert Watcher frequency to pprof period (for time-based
frequency)
* Added general tracepoint support, allowing users to attach to uprobes,
kprobes, and other instrumentable tracing interfaces. This feature is
undocumented, pending support in the Datadog Profiling product.
* Apart from software-based CPU time, all other events now get uploaded as a
tracepoint type

0.8.0 to 0.8.1 (18/03/2022)
* Remove dependency on libstdc++ in library mode

0.7.0 to 0.8.0 (09/03/2022)
* Remove usage of libdwfl internals
* Free inconsistent dwfl modules
We can come into situations where DSOs don't match the dwfl objects.
In this situation we can free all of the PID objects to ensure we will
reload all DSOs and all dwfl objects at the next sample
* Minor fixes for global mode
* Multiple updates in build system
Support ninja, import llvm to a third party directory, compile in c++20,
setup third parties during configure step.
* Library mode 
Deliver profiler as static and shared libraries. DDProf is embedded in the library.
The library daemonizes ddprof. Check out our internal documentation.
* Ensure we export the data gathered at the end of the process's lifetime
This also ensure we export data for short lived processes
* Improved management of signals
SIGTERM and SIGINT now properly end the application
* ARM support

0.6.4 to 0.7.0 (19/01/2022)
* Asynchronous HTTP exports 
- Reduces lost events while exporting
* Support clang builds
* Whole host profiling 
  - Add support to locate files through /proc
* Support url option

0.6.3 to 0.6.4 (06/10/2021)
* Virtual Base frames : Ensure all frames are attached to a base frame
* Explain Unwinding errors through virtual frames (ex: anonymous regions)
* Unwinding fixes (for all processes that fork / global mode)
    - Differentiate dwfl caches per PID
    - Link symbol cache entries to the lifetime of a DSO
* Improve instrumentation sequence : avoids missing initial frames
    - Wait for instrumentation (perf event open) before launching the process
* Build and test different libc versions : added 2.27
* Unwinding fix : Fix check on dwfl_attach_state result
* Support Tags in profiles 
    - Add custom tags through input options
    - Get information on number of CPUs and hw threads in profile tags

0.6.2 to 0.6.3 (28/09/2021)
* Add LTO to build process
* Add _executable flag to DSO cache
* DSO cache is populated with all regions, not just executable ones
* Add PID-by-PID throttling to DSO backpopulate requests, since these requests
  are numerous, expensive, and beyond a single invocation per sample, do not
  improve unwinding quality--the throttle limits backpopulate to the soonest
  of once very 200 samples or once per export
* Fix bug where the lifetime of an mmap()'d region was bound to the lifetime
  of the RegionHolder temporary, causing DSO reads to fail
* Resequence the setuid() mmap maneuver so that an mmap() is attempted with
  current UID before switching
* Filter PID 0 before it hits any of the ddprof worker functions, since we
  currently don't handle it at all
* Fix buggy exit in application start when an invalid argument is given
* Add an overflow check to the unwinding memory_read interface 
* Add some backpopulate and region holder unit tests


src/sha1.c
/*
SHA-1 in C
By Steve Reid <steve@edmweb.com>
100% Public Domain

Test Vectors (from FIPS PUB 180-1)
"abc"
  A9993E36 4706816A BA3E2571 7850C26C 9CD0D89D
"abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq"
  84983E44 1C3BD26E BAAE4AA1 F95129E5 E54670F1
A million repetitions of "a"
  34AA973C D4C4DAA4 F61EEB2B DBAD2731 6534016F
*/

/* #define LITTLE_ENDIAN * This should be #define'd already, if true. */
/* #define SHA1HANDSOFF * Copies data before messing with it. */

#define SHA1HANDSOFF

#include <stdio.h>
#include <string.h>

/* for uint32_t */
#include <stdint.h>

#include "sha1.h"

#define rol(value, bits) (((value) << (bits)) | ((value) >> (32 - (bits))))

/* blk0() and blk() perform the initial expand. */
/* I got the idea of expanding during the round function from SSLeay */
#if BYTE_ORDER == LITTLE_ENDIAN
#  define blk0(i)                                                              \
    (block->l[i] = (rol(block->l[i], 24) & 0xFF00FF00) |                       \
         (rol(block->l[i], 8) & 0x00FF00FF))
#elif BYTE_ORDER == BIG_ENDIAN
#  define blk0(i) block->l[i]
#else
#  error "Endianness not defined!"
#endif
#define blk(i)                                                                 \
  (block->l[i & 15] = rol(block->l[(i + 13) & 15] ^ block->l[(i + 8) & 15] ^   \
                              block->l[(i + 2) & 15] ^ block->l[i & 15],       \
                          1))

/* (R0+R1), R2, R3, R4 are the different operations used in SHA1 */
#define R0(v, w, x, y, z, i)                                                   \
  z += ((w & (x ^ y)) ^ y) + blk0(i) + 0x5A827999 + rol(v, 5);                 \
  w = rol(w, 30);
#define R1(v, w, x, y, z, i)                                                   \
  z += ((w & (x ^ y)) ^ y) + blk(i) + 0x5A827999 + rol(v, 5);                  \
  w = rol(w, 30);
#define R2(v, w, x, y, z, i)                                                   \
  z += (w ^ x ^ y) + blk(i) + 0x6ED9EBA1 + rol(v, 5);                          \
  w = rol(w, 30);
#define R3(v, w, x, y, z, i)                                                   \
  z += (((w | x) & y) | (w & x)) + blk(i) + 0x8F1BBCDC + rol(v, 5);            \
  w = rol(w, 30);
#define R4(v, w, x, y, z, i)                                                   \
  z += (w ^ x ^ y) + blk(i) + 0xCA62C1D6 + rol(v, 5);                          \
  w = rol(w, 30);

/* Hash a single 512-bit block. This is the core of the algorithm. */

void SHA1Transform(uint32_t state[5], const unsigned char buffer[64]) {
  uint32_t a, b, c, d, e;

  typedef struct {
    uint32_t l[16];
  } CHAR64LONG16;

#ifdef SHA1HANDSOFF
  CHAR64LONG16 block[1]; /* use array to appear as a pointer */

  memcpy(block, buffer, 64);
#else
  /* The following had better never be used because it causes the
   * pointer-to-const buffer to be cast into a pointer to non-const.
   * And the result is written through.  I threw a "const" in, hoping
   * this will cause a diagnostic.
   */
  CHAR64LONG16 *block = (const CHAR64LONG16 *)buffer;
#endif
  /* Copy context->state[] to working vars */
  a = state[0];
  b = state[1];
  c = state[2];
  d = state[3];
  e = state[4];
  /* 4 rounds of 20 operations each. Loop unrolled. */
  R0(a, b, c, d, e, 0);
  R0(e, a, b, c, d, 1);
  R0(d, e, a, b, c, 2);
  R0(c, d, e, a, b, 3);
  R0(b, c, d, e, a, 4);
  R0(a, b, c, d, e, 5);
  R0(e, a, b, c, d, 6);
  R0(d, e, a, b, c, 7);
  R0(c, d, e, a, b, 8);
  R0(b, c, d, e, a, 9);
  R0(a, b, c, d, e, 10);
  R0(e, a, b, c, d, 11);
  R0(d, e, a, b, c, 12);
  R0(c, d, e, a, b, 13);
  R0(b, c, d, e, a, 14);
  R0(a, b, c, d, e, 15);
  R1(e, a, b, c, d, 16);
  R1(d, e, a, b, c, 17);
  R1(c, d, e, a, b, 18);
  R1(b, c, d, e, a, 19);
  R2(a, b, c, d, e, 20);
  R2(e, a, b, c, d, 21);
  R2(d, e, a, b, c, 22);
  R2(c, d, e, a, b, 23);
  R2(b, c, d, e, a, 24);
  R2(a, b, c, d, e, 25);
  R2(e, a, b, c, d, 26);
  R2(d, e, a, b, c, 27);
  R2(c, d, e, a, b, 28);
  R2(b, c, d, e, a, 29);
  R2(a, b, c, d, e, 30);
  R2(e, a, b, c, d, 31);
  R2(d, e, a, b, c, 32);
  R2(c, d, e, a, b, 33);
  R2(b, c, d, e, a, 34);
  R2(a, b, c, d, e, 35);
  R2(e, a, b, c, d, 36);
  R2(d, e, a, b, c, 37);
  R2(c, d, e, a, b, 38);
  R2(b, c, d, e, a, 39);
  R3(a, b, c, d, e, 40);
  R3(e, a, b, c, d, 41);
  R3(d, e, a, b, c, 42);
  R3(c, d, e, a, b, 43);
  R3(b, c, d, e, a, 44);
  R3(a, b, c, d, e, 45);
  R3(e, a, b, c, d, 46);
  R3(d, e, a, b, c, 47);
  R3(c, d, e, a, b, 48);
  R3(b, c, d, e, a, 49);
  R3(a, b, c, d, e, 50);
  R3(e, a, b, c, d, 51);
  R3(d, e, a, b, c, 52);
  R3(c, d, e, a, b, 53);
  R3(b, c, d, e, a, 54);
  R3(a, b, c, d, e, 55);
  R3(e, a, b, c, d, 56);
  R3(d, e, a, b, c, 57);
  R3(c, d, e, a, b, 58);
  R3(b, c, d, e, a, 59);
  R4(a, b, c, d, e, 60);
  R4(e, a, b, c, d, 61);
  R4(d, e, a, b, c, 62);
  R4(c, d, e, a, b, 63);
  R4(b, c, d, e, a, 64);
  R4(a, b, c, d, e, 65);
  R4(e, a, b, c, d, 66);
  R4(d, e, a, b, c, 67);
  R4(c, d, e, a, b, 68);
  R4(b, c, d, e, a, 69);
  R4(a, b, c, d, e, 70);
  R4(e, a, b, c, d, 71);
  R4(d, e, a, b, c, 72);
  R4(c, d, e, a, b, 73);
  R4(b, c, d, e, a, 74);
  R4(a, b, c, d, e, 75);
  R4(e, a, b, c, d, 76);
  R4(d, e, a, b, c, 77);
  R4(c, d, e, a, b, 78);
  R4(b, c, d, e, a, 79);
  /* Add the working vars back into context.state[] */
  state[0] += a;
  state[1] += b;
  state[2] += c;
  state[3] += d;
  state[4] += e;
#ifdef SHA1HANDSOFF
  memset(block, '\0', sizeof(block));
#endif
}

/* SHA1Init - Initialize new context */

void SHA1Init(SHA1_CTX *context) {
  /* SHA1 initialization constants */
  context->state[0] = 0x67452301;
  context->state[1] = 0xEFCDAB89;
  context->state[2] = 0x98BADCFE;
  context->state[3] = 0x10325476;
  context->state[4] = 0xC3D2E1F0;
  context->count[0] = context->count[1] = 0;
}

/* Run your data through this. */

void SHA1Update(SHA1_CTX *context, const unsigned char *data, uint32_t len) {
  uint32_t i;

  uint32_t j;

  j = context->count[0];
  if ((context->count[0] += len << 3) < j)
    context->count[1]++;
  context->count[1] += (len >> 29);
  j = (j >> 3) & 63;
  if ((j + len) > 63) {
    memcpy(&context->buffer[j], data, (i = 64 - j));
    SHA1Transform(context->state, context->buffer);
    for (; i + 63 < len; i += 64) {
      SHA1Transform(context->state, &data[i]);
    }
    j = 0;
  } else
    i = 0;
  memcpy(&context->buffer[j], &data[i], len - i);
}

/* Add padding and return the message digest. */

void SHA1Final(unsigned char digest[20], SHA1_CTX *context) {
  unsigned i;

  unsigned char finalcount[8];

  unsigned char c;

#if 0 /* untested "improvement" by DHR */
    /* Convert context->count to a sequence of bytes
     * in finalcount.  Second element first, but
     * big-endian order within element.
     * But we do it all backwards.
     */
    unsigned char *fcp = &finalcount[8];

    for (i = 0; i < 2; i++)
    {
        uint32_t t = context->count[i];

        int j;

        for (j = 0; j < 4; t >>= 8, j++)
            *--fcp = (unsigned char) t}
#else
  for (i = 0; i < 8; i++) {
    finalcount[i] = (unsigned char)((context->count[(i >= 4 ? 0 : 1)] >>
                                     ((3 - (i & 3)) * 8)) &
                                    255); /* Endian independent */
  }
#endif
  c = 0200;
  SHA1Update(context, &c, 1);
  while ((context->count[0] & 504) != 448) {
    c = 0000;
    SHA1Update(context, &c, 1);
  }
  SHA1Update(context, finalcount, 8); /* Should cause a SHA1Transform() */
  for (i = 0; i < 20; i++) {
    digest[i] =
        (unsigned char)((context->state[i >> 2] >> ((3 - (i & 3)) * 8)) & 255);
  }
  /* Wipe variables */
  memset(context, '\0', sizeof(*context));
  memset(&finalcount, '\0', sizeof(finalcount));
}

void SHA1(unsigned char digest[20], const char *str, int len) {
  SHA1_CTX ctx;
  unsigned int ii;

  SHA1Init(&ctx);
  for (ii = 0; ii < len; ii += 1)
    SHA1Update(&ctx, (const unsigned char *)str + ii, 1);
  SHA1Final(digest, &ctx);
}

void SHA1StrDigest(const unsigned char digest[20], char str_digest[41]) {
  for (int i = 0; i < 20; ++i) {
    sprintf(str_digest + 2 * i, "%02x", digest[i]);
  }
}